client/src/components/Multiselect.tsx

```diff
@@ -1,712 +1,348 @@
+  return <Select {...props} />;
+}
 
-  onArrowKeyNavigation(e) {
-    const { options, highlightOption, toggleOptionsList, inputValue, selectedValues } = this.state;
-    const { disablePreSelectedValues } = this.props;
-    if (e.keyCode === 8 && !inputValue && !disablePreSelectedValues && selectedValues.length) {
-      this.onRemoveSelectedItem(selectedValues.length - 1);
-    }
-    if (!options.length) {
-      return;
-    }
-    if (e.keyCode === 38) {
-      if (highlightOption > 0) {
-        this.setState((previousState) => ({
-          highlightOption: previousState.highlightOption - 1,
-        }));
-      } else {
-        this.setState({ highlightOption: options.length - 1 });
+function MultiSelect<
+  Option,
+  IsMulti extends boolean = true,
+  GroupType extends GroupBase<Option> = GroupBase<Option>
+>({
+  options,
+  onApply,
+  infoAlert,
+  onChange,
+  onInputChange,
+  previewSelectedNum,
+  i18n,
+  ...props
+}: Props<Option, IsMulti, GroupType> & CustomMultiselectProps) {
+  const [selections, setSelections] = useState<Option[]>([]);
+  const [hasError, setHasError] = useState(false);
+  const [showAllSelections, setShowAllSelections] = useState(false);
+  const [inputValue, setInputValue] = useState<string | undefined>(undefined);
+
+  const handleChange = useCallback(
+    (newValue, actionMeta) => {
+      if (onChange) onChange(newValue, actionMeta);
+      setSelections(newValue);
+    },
+    [onChange]
+  );
+
+  const handleInputChange = useCallback(
+    (newValue, actionMeta) => {
+      if (onInputChange) onInputChange(newValue, actionMeta);
+      if (!newValue) setHasError(false);
+      if (actionMeta.action !== "input-blur" && actionMeta.action !== "menu-close") {
+        setInputValue(newValue);
       }
-    } else if (e.keyCode === 40) {
-      if (highlightOption < options.length - 1) {
-        this.setState((previousState) => ({
-          highlightOption: previousState.highlightOption + 1,
-        }));
-      } else {
-        this.setState({ highlightOption: 0 });
-      }
-    } else if (e.key === "Enter" && options.length && toggleOptionsList) {
-      if (highlightOption === -1) {
-        return;
-      }
-      this.onSelectItem(options[highlightOption]);
-    }
-  }
-
-  onEscKeyCloseOptionList(e) {
-    if (e.keyCode === KEY_ESC) {
-      e.preventDefault();
-      this.onCloseOptionList();
-      this.searchBox.current.focus();
-    }
-  }
-
-  onRemoveSelectedItem(item) {
-    let { selectedValues, index = 0 } = this.state;
-    const { onRemove, showCheckbox, displayValue, isObject } = this.props;
-    if (isObject) {
-      index = selectedValues.findIndex((i) => i[displayValue] === item[displayValue]);
+    },
+    [onInputChange]
+  );
+
+  // Passing custom prop to make accessible to child compnents via selectProps
+  const removeValue = useCallback(
+    // @ts-ignore (TODO: says property 'value' doesn't exist on type Option)
```

Comment:
hm, can we extend the type Option to include the property `value`? kind of confused about this error

Reply:
Yeah I don't understand it either, because the `Option` type has value, and it knows it has this type, but for some reason it's doesn't know what the properties are. 

![image](https://user-images.githubusercontent.com/16906516/231182141-dff33977-18d6-43f9-97ab-31489c133190.png)

At the top of this files it's defined as:

```
export type Option = {
  value: string;
  label: string;
  [key: string]: string;
};
```